File: Prácticas/P4/JAVA/Civitas/src/civitas/JugadorEspeculador.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package civitas;

/**
 *
 * @author Jose Padial
 */
public class JugadorEspeculador extends Jugador{
    private int fianza;
    protected static final int factorEspeculador = 2;
    
    public JugadorEspeculador(Jugador otro, int fianza) {
        super(otro);
        this.fianza = fianza;
    }
    
   private boolean puedoEdificarCasa(TituloPropiedad propiedad){
        boolean puede = false;
        
        for(int i = 0; i < propiedades.size(); i++){
            if(propiedad == propiedades.get(i) && propiedades.get(i).getNumCasas() < casasMax*factorEspeculador){
                puede = true;
            }
        }
        
        return puede;
    }
    
    private boolean puedoEdificarHotel(TituloPropiedad propiedad){
        boolean puede = false;
        
        for(int i = 0; i < propiedades.size(); i++){
            if(propiedad == propiedades.get(i) && propiedades.get(i).getNumHoteles() < hotelesMax*factorEspeculador){
                if(propiedad.getNumCasas()>=getCasasPorHotel()){
                puede = true;
                }
            }
        }
        
        return puede;
    }
    
    @Override
    boolean pagaImpuesto(float cantidad){
        if(encarcelado == true){
            return false;
        }
        else{
            return paga(cantidad/2);
        }
    }
    
    @Override
    boolean encarcelar(int numCasillaCarcel){
        if(debeSerEncarcelado() == true && !paga(precioLibertad)){
            moverACasilla(numCasillaCarcel);
            encarcelado = true;
            diario.ocurreEvento("Jugador " + super.getNombre() + " encarcelado");
            return encarcelado;
        }
        else{
            return encarcelado;
        }
        
    }

    @Override
    public String toString() {
        return  "JugadorEspeculador{" + super.toString() + "fianza=" + fianza + '}';
    }
    
    
    
}

File: Prácticas/P4/RUBY/Civitas/lib/jugador_especulador.rb
# To change this license header, choose License Headers in Project Properties.
# To change this template file, choose Tools | Templates
# and open the template in the editor.

module Civitas
  class JugadorEspeculador < Jugador
    @@FACTOR_ESPECULADOR = 2
    def initialize(otro, fianza)
      super(otro)
      @fianza = fianza
    end
    
    def puedo_edificar_casa(propiedad)
      puede = false
      
      for i in @propiedades
        if(propiedad == @propiedades[i] && @propiedades[i].get_num_casas < @@CASAS_MAX*@@FACTOR_ESPECULADOR)
          puede = true
        end
      end
      puede
    end
    
    def puedo_edificar_hotel(propiedad)
      puede = false
      
      for i in @propiedades
        if(propiedad == @propiedades[i] && @propiedades[i].get_num_hoteles < @@HOTELES_MAX*@@FACTOR_ESPECULADOR)
          puede = true
        end
      end
      puede
    end
    
    def paga_impuesto(cantidad)
      if(@encarcelado == true)
          return false
      else
          return paga(cantidad/2)
      end  
    end
    
    def encarcelar(num_casilla_carcel)
      if(debe_ser_encarcelado() == true && !paga(@@PRECIO_LIBERTAD))
            mover_a_casilla(num_casilla_carcel)
            @encarcelado = true
            @diario.ocurre_evento("Jugador " + @nombre + " encarcelado")
            return @encarcelado
      else
            return @encarcelado
      end
    end
    
    def to_s
      to_s = "Jugador Especulador Fianza #{@fianza} #{super}"
    end
  end
end
